rust/arrow/src/array/array_list.rs

```diff
@@ -139,7 +166,19 @@ impl<OffsetSize: OffsetSizeTrait> From<ArrayDataRef> for GenericListArray<Offset
             1,
             "ListArray should contain a single child array (values array)"
         );
-        let values = make_array(data.child_data()[0].clone());
+
+        let values = data.child_data()[0].clone();
+
+        if let Some(child) = Self::get_type(data.data_type()) {
```

Comment:
can we have a few tests to cover this? checking error message and all. 

Also I'm not sure if `assert_eq` is good here:  IMO assertion should only be used for checking internal logic that developer should follow and which are not exposed to the library users, but in this case it appears not. It's just a nit though since this is already used in multiple places before.

Reply:
I agree with you. However, that requires a larger change as we would need to move from `From` to `TryFrom`, so for now I just want to avoid `unsafe` code by panicking everytime something may go wrong.